src/traffic_server/InkIOCoreAPI.cc

```diff
@@ -253,9 +254,11 @@ void
 TSMutexDestroy(TSMutex m)
 {
   sdk_assert(sdk_sanity_check_mutex(m) == TS_SUCCESS);
-  ink_release_assert(((ProxyMutex *)m)->refcount() == 0);
+  ProxyMutex *mutexp = reinterpret_cast<ProxyMutex *>(m);
+  mutexp->refcount_dec();
+  ink_release_assert((mutexp)->refcount() == 0);
```

Comment:
I wonder if we really need this - might it not be reasonable to enable the normal reference counting to work here? In fact, this might obviate the whole point of doing the reference couting in `TSMutexLock`, the purpose of which is to preserve the mutex over destruction by the plugin. That seems like to assert now.

Reply:
Talked with @SolidWallOfCode and indeed, we should not explicitly free here.  Rather just decrement the ref count and wait for the smart pointer logic go to 0 and free.  If there is some other element of the code holding a smart pointer at this point, it would be wrong to explicitly free here.